File: nerarith-haskell/day4/4a.hs
import Data.List.Split (splitOn)
import qualified Data.Map as Map

main = interact (output . map (isValid . readInput) . splitOn [""] . lines)

readInput :: [String] -> Map.Map String String
readInput = Map.fromList . map splitPair . concatMap words

splitPair :: String -> (String, String)
splitPair s = (key, value)
    where (key:value:_) = splitOn ":" s

isValid :: Map.Map String String -> Bool
isValid d = and $ map (flip Map.member d) ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

output :: [Bool] -> String
output = (++"\n") . show . length . filter (True==)

File: cernael-python/4a.py
import re
def solve(lines):
    passports = []
    passport = {}
    for i in range(len(lines)):
        if ord(lines[i][0]) == 10:
            passports.append(passport)
            passport = {}
        else:
            fields = lines[i].split()
            for f in range(len(fields)):
                (k,v) = fields[f].split(':')
                passport[k] = v
    passports.append(passport)

    valid = 0
    for i in range(len(passports)):
        p = passports[i]
        if ('byr' in p and
            'iyr' in p and
            'eyr' in p and
            'hgt' in p and
            'hcl' in p and
            'ecl' in p and
            'pid' in p
        ):
            valid += 1
    return valid
if __name__ == '__main__':
    lines = []
    with open('4.txt') as f:
        for line in f.readlines():
            lines.append(line)
    print(solve(lines))
